File: benches/stats_suite/cpp/five_number_summary.cpp
// BM-001: Five-Number Summary Implementation (C++)
// Reference implementation for cross-language benchmarking
//
// Computes: min, Q1, median, Q3, max, IQR, and outlier count
// Input: Whitespace-separated numbers from stdin
// Output: Statistical summary to stdout

#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <iomanip>
#include <cmath>

using namespace std;
using namespace std::chrono;

// TODO: Implement five-number summary calculation
// This is a placeholder for the actual implementation

struct Statistics {
    double min;
    double q1;
    double median;
    double q3;
    double max;
    double iqr;
    int outlier_count;
    double outlier_percentage;
};

Statistics compute_five_number_summary(vector<double>& data) {
    Statistics stats;
    
    if (data.empty()) {
        cerr << "Error: Empty dataset" << endl;
        exit(1);
    }
    
    // Sort the data (required for quantile calculation)
    sort(data.begin(), data.end());
    
    size_t n = data.size();
    
    // TODO: Implement actual calculation
    // For now, this is a stub that returns placeholder values
    stats.min = data[0];
    stats.max = data[n - 1];
    stats.median = 0.0;  // Placeholder
    stats.q1 = 0.0;      // Placeholder
    stats.q3 = 0.0;      // Placeholder
    stats.iqr = 0.0;     // Placeholder
    stats.outlier_count = 0;
    stats.outlier_percentage = 0.0;
    
    return stats;
}

int main() {
    // Read input data
    vector<double> data;
    double value;
    
    while (cin >> value) {
        data.push_back(value);
    }
    
    if (data.empty()) {
        cerr << "Error: No data read from input" << endl;
        return 1;
    }
    
    cout << "Read " << data.size() << " data points" << endl;
    
    // Start timing (exclude I/O time)
    auto start = high_resolution_clock::now();
    
    Statistics stats = compute_five_number_summary(data);
    
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end - start);
    
    // Output results
    cout << fixed << setprecision(3);
    cout << "Min: " << stats.min << endl;
    cout << "Q1: " << stats.q1 << endl;
    cout << "Median: " << stats.median << endl;
    cout << "Q3: " << stats.q3 << endl;
    cout << "Max: " << stats.max << endl;
    cout << "IQR: " << stats.iqr << endl;
    cout << "Outliers: " << stats.outlier_count 
         << " (" << setprecision(2) << stats.outlier_percentage << "%)" << endl;
    cout << "\nExecution Time: " << duration.count() << " microseconds" << endl;
    
    return 0;
}

File: benches/stats_suite/rust/five_number_summary.rs
// BM-001: Five-Number Summary Implementation (Rust)
// Cross-language benchmark comparison with C++ and FlameLang
//
// Computes: min, Q1, median, Q3, max, IQR, and outlier count
// Input: Whitespace-separated numbers from stdin
// Output: Statistical summary to stdout

use std::io::{self, BufRead};
use std::time::Instant;

#[derive(Debug)]
struct Statistics {
    min: f64,
    q1: f64,
    median: f64,
    q3: f64,
    max: f64,
    iqr: f64,
    outlier_count: usize,
    outlier_percentage: f64,
}

// TODO: Implement five-number summary calculation
// This is a placeholder for the actual implementation
fn compute_five_number_summary(data: &mut Vec<f64>) -> Statistics {
    if data.is_empty() {
        eprintln!("Error: Empty dataset");
        std::process::exit(1);
    }
    
    // Sort the data (required for quantile calculation)
    data.sort_by(|a, b| a.partial_cmp(b).unwrap());
    
    let n = data.len();
    
    // TODO: Implement actual calculation
    // For now, this is a stub that returns placeholder values
    Statistics {
        min: data[0],
        max: data[n - 1],
        median: 0.0,  // Placeholder
        q1: 0.0,      // Placeholder
        q3: 0.0,      // Placeholder
        iqr: 0.0,     // Placeholder
        outlier_count: 0,
        outlier_percentage: 0.0,
    }
}

fn main() {
    // Read input data
    let stdin = io::stdin();
    let mut data = Vec::new();
    
    for line in stdin.lock().lines() {
        if let Ok(line) = line {
            for token in line.split_whitespace() {
                if let Ok(value) = token.parse::<f64>() {
                    data.push(value);
                }
            }
        }
    }
    
    if data.is_empty() {
        eprintln!("Error: No data read from input");
        std::process::exit(1);
    }
    
    println!("Read {} data points", data.len());
    
    // Start timing (exclude I/O time)
    let start = Instant::now();
    
    let stats = compute_five_number_summary(&mut data);
    
    let duration = start.elapsed();
    
    // Output results
    println!("Min: {:.3}", stats.min);
    println!("Q1: {:.3}", stats.q1);
    println!("Median: {:.3}", stats.median);
    println!("Q3: {:.3}", stats.q3);
    println!("Max: {:.3}", stats.max);
    println!("IQR: {:.3}", stats.iqr);
    println!("Outliers: {} ({:.2}%)", stats.outlier_count, stats.outlier_percentage);
    println!("\nExecution Time: {} microseconds", duration.as_micros());
}
